### PR TITLE
feat(dateRangePicker): Adding DateRangePicker feature

### PR DESCRIPTION
**Pull request by mustafapsd**
Hi! I've started to work on date range picker feature https://github.com/vlio20/angular-datepicker/issues/54. I'm a bit confused, I hope I'm doing correct. Features to be added:

1. Hover indication of the range.
2. Max & MIn range validations.
3. Preview format.
4. inline mode.
5. Multiple ranges (advanced feature - not mandatory).

### CODE REVIEW

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
-import {DayCalendarService} from '../day-calendar/day-calendar.service';
-import {IDayTimeCalendarConfig} from '../day-time-calendar/day-time-calendar-config.model';
-import {DayTimeCalendarService} from '../day-time-calendar/day-time-calendar.service';
-import {ITimeSelectConfig} from '../time-select/time-select-config.model';
-import {TimeSelectComponent} from '../time-select/time-select.component';
-import {TimeSelectService} from '../time-select/time-select.service';
-import {IDatePickerConfig, IDatePickerConfigInternal} from './date-picker-config.model';
-import {IDpDayPickerApi} from './date-picker.api';
-import {DatePickerService} from './date-picker.service';
+import { IDate, IDateRange } from '../common/models/date.model';
```
Comment (vlio20) [inline projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:1]: please revet those changes as they are not relevant

File: projects/ng2-date-picker/.eslintrc.json [hunk 2]
```diff
@@ -10,8 +10,8 @@
       ],
       "parserOptions": {
         "project": [
-          "projects/ng2-datepicker/tsconfig.lib.json",
-          "projects/ng2-datepicker/tsconfig.spec.json"
+          "projects/ng2-date-picker/tsconfig.lib.json",
```
Comment (vlio20) [inline projects/ng2-date-picker/.eslintrc.json:13]: how is that relevant to your changes? I think that you pulled from older version and not from the one in the master

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts [hunk 3]
```diff
@@ -78,10 +78,10 @@ import {ConnectionPositionPair} from '@angular/cdk/overlay';
   ]
 })
 export class DatePickerComponent implements OnChanges,
-                                            OnInit,
-                                            ControlValueAccessor,
-                                            Validator,
-                                            OnDestroy {
+  OnInit,
```
Comment (vlio20) [inline projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:81]: Please revert all linting changes.

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts [hunk 4]
```diff
@@ -93,13 +93,16 @@ export class DatePickerComponent implements OnChanges,
   @Input() maxDate: SingleCalendarValue;
   @Input() minTime: SingleCalendarValue;
   @Input() maxTime: SingleCalendarValue;
+  @Input() dateRangePicker: boolean = false;
+  @Input() dateRange: IDateRange;
```
Comment (vlio20) [inline projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:97]: The range value should be set through the controller input - not as an additional Input of the component.

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts [hunk 5]
```diff
@@ -384,6 +392,16 @@ export class DatePickerComponent implements OnChanges,
     });
   }
 
+  rangeSelected(range: IDateRange, ignoreClose?: boolean) {
+    if (!ignoreClose) {
+      this.onDateClick();
+    }
+
+    this.inputElementValue = (this.utilsService.convertToString(range.from, this.componentConfig.format))
+      + ' <> ' + (this.utilsService.convertToString(range.to, this.componentConfig.format));
```
Comment (vlio20) [inline projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:401]: 1. Please use template strings
2. Let's make the delimiter to be configured

File: projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.less [hunk 6]
```diff
@@ -43,12 +43,32 @@
       color: @c-white;
     }
 
-    .dp-prev-month, .dp-next-month {
+    .dp-start-date {
+      background: @c-primary !important;
```
Comment (vlio20) [inline projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.less:47]: please avoid using important. add additional classes if needed.

File: projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.ts [hunk 7]
```diff
@@ -219,6 +225,31 @@ export class DayCalendarComponent implements OnInit, OnChanges, ControlValueAcce
...
+      }
+
+      if (day.date.isBefore(this.dateRange.from)) {
+        this.dateRange.from = day.date;
+        return;
+      }
+
+      this.dateRange.to = day.date;
+      this.onDateRangeSelect.emit(this.dateRange);
+      // eslint-disable-next-line no-console
```
Comment (vlio20) [inline projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.ts:247]: please remove

**Review by vlio20 — CHANGES_REQUESTED** [review]

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts (hunk 3 above)
Comment (mustafapsd) [reply projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:81]: Opps, sorry

File: projects/ng2-date-picker/.eslintrc.json (hunk 2 above)
Comment (mustafapsd) [reply projects/ng2-date-picker/.eslintrc.json:13]: Imports are wrong, so the editor gives errors. You can see these wrong imports.

https://github.com/vlio20/angular-datepicker/blob/master/projects/ng2-date-picker/.eslintrc.json

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts (hunk 4 above)
Comment (mustafapsd) [reply projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:97]: I thought that user can be set default values. This is not work now. Should I remove this?

File: projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.ts (hunk 7 above)
Comment (mustafapsd) [reply projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.ts:247]: What's wrong here? Thank you for review.

Comment (mustafapsd) [reply projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.ts:247]: Oh okey, you want to remove eslint-disable line. I currently using this line for testing.

File: projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts (hunk 5 above)
Comment (mustafapsd) [reply projects/ng2-date-picker/src/lib/date-picker/date-picker.component.ts:401]: 1. Ok
2. I'll work for limiters tomorrow.

File: projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.less (hunk 6 above)
Comment (mustafapsd) [reply projects/ng2-date-picker/src/lib/day-calendar/day-calendar.component.less:47]: I've removed, thanks

**Comment by mustafapsd** [comment]
I will work tomorrow on the features that you wrote. I did the changes that you mentioned in the review. Thank you for your review, this is really important for me! Could you share a formatting config file (like `.prettierrc`)?

**Comment by vlio20** [comment]
@mustafapsd I don't have one. I am ok with adding a prettier but it should be on a different PR.

**Comment by mustafapsd** [comment]
I added hover effect and inline mode. 

[Video](https://www.screencast.com/t/yEB8L8dX9)

**Comment by vlio20** [comment]
@mustafapsd it's very hard to follow the changes because of the linters. Can you please undo them so I can review the actual code that is related to the range feature? 
 
Few more notes:
1. Maybe it is better to have a dedicated component for the range picker?
2. Other calendars should also be supported - dateTime/month. 

We would also need to add e2e tests.

**Comment by mustafapsd** [comment]
Sorry for linters, writing without formatting is really hard :S 

Actually I think the same thing. This will be getting bigger day to day, so will be hard to change. 

I never saw time or month pickers in date range pickers. Is it really necessary?

**Comment by mustafapsd** [comment]
I removed linting changes for now. Is this logic true for date range pickers? If I'm doing it correctly then I'll create a new component for the range picker.

**Comment by vlio20** [comment]
I will try to pull your fork and play with it and give you my feedback

**Comment by mustafapsd** [comment]
Thank you!
